zk/stages/stage_witness.go

```diff
@@ -0,0 +1,547 @@
+		if err != nil {
+			return fmt.Errorf("failed to open tx, %w", err)
+		}
+		defer tx.Rollback()
+	}
+
+	// batch until which witness has been generated and saved to db
+	witnessProgress, err := stages.GetStageProgress(tx, stages.Witness)
+	if err != nil {
+		return fmt.Errorf("save stage progress error: %v", err)
+	}
+
+	// batch which has been processed by stage_batches
+	latestBlock, err := stages.GetStageProgress(tx, stages.Execution)
+	if err != nil {
+		return fmt.Errorf("could not retrieve batch no progress")
+	}
+
+	hermezDbReader := hermez_db.NewHermezDbReader(tx)
+
+	batchesProgress, err := hermezDbReader.GetBatchNoByL2Block(latestBlock)
+	if err != nil {
+		return fmt.Errorf("could not retrieve batch for latest block")
+	}
+
+	if witnessProgress >= batchesProgress {
+		log.Info("All batch witness already in db, skipping")
+		return nil
+	}
+
+	// witness generation should start from stored witness of batch in db + 1
+	witnessFromBatch := witnessProgress + 1
+	// witness generation should happen till batch of latest block -1
+	// as there might be more blocks incoming for this batch
+	witnessToBatch := batchesProgress - 1
+
+	// Don't generate witness for batches that will eventually be deleted in
+	// the same cycle
+	if witnessFromBatch+cfg.storeNBatches < witnessToBatch {
+		witnessFromBatch = witnessToBatch - cfg.storeNBatches + 1
+	}
+
+	wg := NewWitnessGenerator(cfg.tmpDir, cfg.historyV3, cfg.agg, cfg.blockReader, cfg.chainConfig, cfg.engine)
+
+	for b := witnessFromBatch; b <= witnessToBatch; b++ {
+		highestBlockInBatch, err := hermezDbReader.GetHighestBlockInBatch(b)
+		if err != nil {
+			return err
+		}
+
+		if highestBlockInBatch == 0 {
+			continue
+		}
+
+		nextStageProgress, err := stages.GetStageProgress(tx, stages.HashState)
+		if err != nil {
+			return err
+		}
+		if nextStageProgress < highestBlockInBatch {
+			// Skip this stage
+			return nil
+		}
+
+		lowestBlockInBatch, err := hermezDbReader.GetLowestBlockInBatch(b)
+		if err != nil {
+			return err
+		}
+
+		// Skip witness generation for this batch as it is bound to fail
+		if highestBlockInBatch-lowestBlockInBatch > maxGetProofRewindBlockCount {
+			continue
+		}
+
+		log.Info(fmt.Sprintf("[%s] Starting witness generation", logPrefix), "currentBatch", b, "highestBatch", witnessToBatch)
+		w, err := wg.GenerateWitnessByBatch(tx, ctx, b, false)
+		if err != nil {
+			return err
+		}
+
+		hermezDb := hermez_db.NewHermezDb(tx)
+		err = hermezDb.WriteWitnessByBatchNo(b, w)
+		if err != nil {
+			return err
+		}
+
+		oldestBatch, err := hermezDb.GetOldestStoredWitnessBatch()
+		if err != nil {
+			return err
+		}
+
+		// Only storeNBatches witness in db
+		// Won't enforce a hard count as batches aren't contiguous
+		// Cannot use cursor count to get the exact number as batch
+		// witnesses are stored in chunks (number will vary from batch to batch)
+		if (b - oldestBatch + 1) > cfg.storeNBatches {
+			hermezDb.DeleteWitnessByBatchNo(oldestBatch)
+		}
```

Comment:
When using batch number, we can't control the exact max number of witnesses stored in db, because batch number is not contiguous as you mentioned. If we use block number instead, we will have a more accurate control of how many blocks' witnesses we want to keep in the db. Basically, changing `storeNBatches` to `storeNBlocks`, and check if the batch number of the lower bound block (`currentBlockN - storeNBatches - 1`) is stored in db. If so, delete the batch.